### PR TITLE
Add NFT templete and UT

### LINKED ISSUE #239: Write an NFT contract example
> We also need an implementation with decimals greater than zero.  [NFT standard](https://github.com/neo-project/proposals/pull/41)

Some examples:
1. [NEP-11 python demo](https://gist.github.com/hal0x2328/3237fd9f61132cea6b6dd43899947753), _posted by_  @hal0x2328
2. [RyuNFTs-NEO demo](https://github.com/RyuGames/RyuNFTs-NEO)  _posted by_ @WyattMufson 
3. [Zoro-Contract demo](https://github.com/ZoroChain/Zoro-Contracts/blob/master/NFT/NFTContract/NFTContract.cs) _posted by_ @ProDog

Don't forget copy the demo to https://github.com/neo-project/examples/issues/30

### PR DESCRIPTION
**Pull request by ShawnYun**
close #239 
and copy the demo to neo-project/examples#30

### CODE REVIEW

File: templates/Template.NFT.CSharp/NFTContract.cs [hunk 1]
```diff
@@ -0,0 +1,160 @@
...
+    public class NFTContract : SmartContract
+    {
+        [DisplayName("MintedToken")]
+        public static event Action<byte[], byte[], byte[]> MintedToken;//(byte[] to, byte[] tokenId, byte[] properties);
+
+        [DisplayName("Transferred")]
+        public static event Action<byte[], byte[], BigInteger, byte[]> Transferred; //(byte[] from , byte[] to, BigInteger amount, byte[] TokenId)
+
+        //super admin address
+        private static readonly byte[] superAdmin = Helper.ToScriptHash("Nj9Epc1x2sDmd6yH5qJPYwXRqSRf5X6KHE");
```
Comment (vncoelho) [inline templates/Template.NFT.CSharp/NFTContract.cs:22]: Is there possibility to change `superAdmin`?

Comment (ShawnYun) [reply templates/Template.NFT.CSharp/NFTContract.cs:22]: I think it's better not to change it.

**Comment by ShawnYun** [comment]
@vncoelho Thanks. and there are some UTs need to modify neo.

File: templates/Template.NFT.CSharp/NFTContract.cs [hunk 2]
```diff
@@ -0,0 +1,166 @@
...
+        private static string TotalSupplyKey() => "totalSupply";
+
+        private static StorageContext Context() => Storage.CurrentContext;
+        private static byte[] TokenOwnerKey(byte[] tokenId, byte[] owner) => new byte[] { 0x10 }.Concat(tokenId).Concat(owner);
```
Comment (Tommo-L) [inline templates/Template.NFT.CSharp/NFTContract.cs:27]:
```suggestion
        private static byte[] TokenOwnerKey(byte[] tokenId) => new byte[] { 0x10 }.Concat(tokenId);
```

File: templates/Template.NFT.CSharp/NFTContract.cs [hunk 3]
```diff
@@ -0,0 +1,166 @@
...
+        private static byte[] TotalBalanceKey(byte[] owner) => new byte[] { 0x11 }.Concat(owner);
+        private static byte[] TokenBalanceKey(byte[] owner, byte[] tokenId) => new byte[] { 0x12 }.Concat(owner).Concat(tokenId);
```
Comment (Tommo-L) [inline templates/Template.NFT.CSharp/NFTContract.cs:29]:
```suggestion
        private static byte[] TokenBalanceKey(byte[] owner) => new byte[] { 0x12 }.Concat(owner);
```

File: src/Neo.SmartContract.Framework/Helper.cs [hunk 4]
```diff
@@ -148,6 +148,17 @@ public static byte ToByte(this int source)
             return (byte)(source + 0);
         }
 
+        public static bool Equals(this byte[] left, byte[] right)
+        {
+            if (left.Length != right.Length) return false;
+            if (left == null || right == null) return false;
```
Comment (shargon) [inline src/Neo.SmartContract.Framework/Helper.cs:154]: This condition should be before line 153

File: templates/Template.NFT.CSharp/NFTContract.cs [hunk 5]
```diff
@@ -0,0 +1,166 @@
...
+
+namespace NFTContract
+{
+    /// <summary>
+    /// Non-Fungible Token Smart Contract Template
+    /// </summary>
+    public class NFTContract : SmartContract
+    {
+        [DisplayName("MintedToken")]
+        public static event Action<byte[], byte[], byte[]> MintedToken;//(byte[] to, byte[] tokenId, byte[] properties);
```
Comment (shargon) [inline templates/Template.NFT.CSharp/NFTContract.cs:17]: Remove this comments?

File: templates/Template.NFT.CSharp/NFTContract.cs [hunk 6]
```diff
@@ -0,0 +1,166 @@
...
+        }
+
+        public static string Symbol()
+        {
+            return "MNFT";
+        }
+
+        public static string SupportedStandards()
+        {
+            return "{\"NEP-10\", \"NEP-11\"}";
```
Comment (shargon) [inline templates/Template.NFT.CSharp/NFTContract.cs:48]: It should be an array of strings, not a json

File: templates/Template.NFT.CSharp/NFTContract.cs [hunk 7]
```diff
@@ -0,0 +1,166 @@
...
+        public static bool MintNFT(byte[] tokenId, byte[] owner, byte[] properties)
+        {
+            if (!Runtime.CheckWitness(superAdmin)) return false;
+
+            if (owner.Length != 20) throw new FormatException("The parameter 'owner' should be 20-byte address."); ;
+            if (properties.Length > 2048) throw new FormatException("The length of 'properties' should be less than 2048."); ;
+
+            if (Storage.Get(Context(), TokenOwnerKey(tokenId, owner)) != null) return false;
+
+            Storage.Put(Context(), PropertiesKey(tokenId), properties);
```
Comment (shargon) [inline templates/Template.NFT.CSharp/NFTContract.cs:86]: Could you use StorageMap?

Comment (ShawnYun) [reply templates/Template.NFT.CSharp/NFTContract.cs:86]: Yes, I will modify it.

File: templates/Template.NFT.CSharp/NFTContract.cs [hunk 8]
```diff
@@ -0,0 +1,171 @@
...
+        public static event Action<byte[], byte[], byte[]> MintedToken;
+
+        [DisplayName("Transferred")]
+        public static event Action<byte[], byte[], BigInteger, byte[]> Transferred;
+
+        private static readonly byte[] superAdmin = Helper.ToScriptHash("Nj9Epc1x2sDmd6yH5qJPYwXRqSRf5X6KHE");
+        private static string TotalSupplyKey() => "totalSupply";
+
+        private static StorageContext Context() => Storage.CurrentContext;
+        private static byte[] StorageProfixTokenOwner(byte[] tokenId) => new byte[] { 0x10 }.Concat(tokenId);
```
Comment (joeqian10) [inline templates/Template.NFT.CSharp/NFTContract.cs:26]: Change "StorageProfixTokenOwner" to "StoragePrefixTokenOwner"?

Comment (ShawnYun) [reply templates/Template.NFT.CSharp/NFTContract.cs]: Thanks.

File: templates/Template.NFT.CSharp/NFTContract.cs [hunk 9]
```diff
@@ -0,0 +1,194 @@
...
+        }
+
+        public static string Symbol()
+        {
+            return "MNFT";
+        }
+
+        public static string SupportedStandards()
+        {
+            return "NEP-10, NEP-11";
```
Comment (shargon) [inline templates/Template.NFT.CSharp/NFTContract.cs:46]: https://github.com/neo-project/proposals/blob/master/nep-10.mediawiki#supportedstandards array of strings

Comment (ShawnYun) [reply templates/Template.NFT.CSharp/NFTContract.cs]: Done.

File: templates/Template.NFT.CSharp/NFTContract.cs [hunk 10]
```diff
@@ -0,0 +1,194 @@
...
+{
+    /// <summary>
+    /// Non-Fungible Token Smart Contract Template
+    /// </summary>
+    public class NFTContract : SmartContract
+    {
+        [DisplayName("MintedToken")]
+        public static event Action<byte[], byte[], byte[]> MintedToken;
+
+        [DisplayName("Transferred")]
```
Comment (shargon) [inline templates/Template.NFT.CSharp/NFTContract.cs:18]: Why not `Transfer`?

File: templates/Template.NFT.CSharp/NFTContract.cs [hunk 11]
```diff
@@ -0,0 +1,194 @@
...
+        private const byte Prefix_TokenBalance = 12;
+        private const byte Prefix_Properties = 13;
+        private const byte Prefix_TokensOf = 14;
+
+        private const int TOKEN_DECIMALS = 8;
+        private const int FACTOR = 100_000_000;
+
+        public static string Name()
+        {
+            return "MyNFT";
```
Comment (shargon) [inline templates/Template.NFT.CSharp/NFTContract.cs:36]:
```suggestion
            return "NFT";
```

File: templates/Template.NFT.CSharp/NFTContract.cs (hunk 1 above)
Comment (devhawk) [reply templates/Template.NFT.CSharp/NFTContract.cs:24]: contractOwner?

File: templates/Template.NFT.CSharp/NFTContract.cs [hunk 12]
```diff
@@ -0,0 +1,219 @@
...
+
+        private static readonly byte[] superAdmin = Helper.ToScriptHash("Nj9Epc1x2sDmd6yH5qJPYwXRqSRf5X6KHE");
+
+        private static StorageContext Context() => Storage.CurrentContext;
+
+        private const byte Prefix_TotalSupply = 10;
+        private const byte Prefix_TokenOwner = 11;
+        private const byte Prefix_TokenBalance = 12;
+        private const byte Prefix_Properties = 13;
+        private const byte Prefix_TokensOf = 14;
```
Comment (igormcoelho) [inline templates/Template.NFT.CSharp/NFTContract.cs:32]: I think it's quite dangerous to keep using "byte" as prefix marker... it's much better to use "sbyte" instead. I mean, this code may suddently break if some unexpected user decides to put values over 127... Better incentive better practices, I think. Do you agree @erikzhang ?

**Comment by erikzhang** [comment]
This should be moved to https://github.com/neo-project/examples.

**Pull request closed without merging**